src/iface/fragmentation.rs

```diff
@@ -71,12 +71,84 @@ impl<'a> PacketAssembler<'a> {
             total_size,
             last_updated: start_time,
             started_on: start_time,
+            ttl,
             offset_correction,
         };
 
         Ok(())
     }
 
+    /// Return the TTL of the packet.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerNotInit`] when the assembler was not initialized (try initializing the
+    /// assembler with [Self::start]).
+    pub(crate) fn ttl(&self) -> Result<Duration> {
+        match self.assembler {
+            AssemblerState::NotInit => Err(Error::PacketAssemblerNotInit),
+            AssemblerState::Assembling { ttl, .. } => Ok(ttl),
+        }
+    }
+
+    /// Update the TTL of the packet.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerNotInit`] when the assembler was not initialized (try initializing the
+    /// assembler with [Self::start]).
+    pub(crate) fn update_ttl(&mut self, new_ttl: Duration) -> Result<()> {
+        match self.assembler {
+            AssemblerState::NotInit => Err(Error::PacketAssemblerNotInit),
+            AssemblerState::Assembling { ref mut ttl, .. } => {
+                *ttl = new_ttl;
+                Ok(())
+            }
+        }
+    }
+
+    /// Add a hole with a specified size at some specified position with an offset.
+    /// This will modify the assembler and the check if the buffer has enough space.
+    ///
+    /// # Note
+    ///
+    /// This can only work correctly when the offset used when adding data to the buffer is already
+    /// anticipating the extending of the buffer. For example, the offset in the IPv4 header is the
+    /// offset in the total packet. However, when receiving the first packet, we don't know yet
+    /// what the total size will be, but the offset is correct.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerNotInit`] when the assembler was not initialized (try initializing the
+    /// assembler with [Self::start]).
+    /// - Returns [`Error::PacketAssemblerBufferTooSmall`] when the buffer is too small.
+    pub(crate) fn extend_at(&mut self, offset: usize, size: usize) -> Result<()> {
```

Comment:
I'm not sure what `extend_at` is supposed to do, but I think it won't behave correctly if yuo receive the same fragment twice, as it'll do `*total_size += size` twice.

I think it's possible to simplify it:

- The assembler has
   - `total_size: Option<usize>`, so that the total size  *might be unknown* (this is the new bit)
   - Packet data
- On fragment receive:
   - Add the data
   - If last fragment (`more_frags=0`) then we learned total_size, so set it.
- Assembler is complete when
   - total_size is known, AND
   - 0..total_size is covered by data.